File: examples/upstream-demo-app-ios/AstronomyShop/Shop/ShopTelemetry.swift
import Foundation
import OpenTelemetryApi
import OpenTelemetrySdk

/// Single source of truth for every log, span, and metric the demo emits.
///
/// Three different call sites exercise the shop:
///   - `CartViewModel` for user-driven cart actions
///   - `RootState` / `ProductDetailView` for product-view on appear
///   - `AutoDemoDriver` for the `DASH0_AUTO_DEMO=1` loop
///
/// Without this helper they would each reproduce the same 14-span checkout
/// tree, 3-span browse tree, metric instrument setup, and log schema —
/// guaranteed drift. Routing everything through `ShopTelemetry` means
/// renaming a span or adding an attribute happens in exactly one place.
///
/// The class is `@unchecked Sendable` / not actor-isolated so callers on any
/// thread/queue (MainActor UI code, the auto-demo's background
/// `DispatchSourceTimer`) can share a single instrument graph — OTel-Swift's
/// `LongCounter` / `DoubleHistogram` / `Tracer` / `Logger` are documented as
/// thread-safe.
final class ShopTelemetry: @unchecked Sendable {
    let tracer: Tracer
    let logger: Logger
    private var itemsAddedCounter: LongCounter
    private var checkoutDurationHistogram: DoubleHistogram
    private var viewProductHistogram: DoubleHistogram

    /// Build from an `OTelMobile` instance. Returns nil if the SDK isn't
    /// started or any required handle is missing — the UI still renders in
    /// that case, it just emits nothing.
    init?(tracer: Tracer?, logger: Logger?, meter: MeterSdk?) {
        guard let tracer = tracer, let logger = logger, let meter = meter else {
            return nil
        }
        self.tracer = tracer
        self.logger = logger
        self.itemsAddedCounter = meter.counterBuilder(name: "shop.cart.items_added").build()
        self.checkoutDurationHistogram = meter.histogramBuilder(name: "shop.checkout.duration_ms").build()
        self.viewProductHistogram = meter.histogramBuilder(name: "shop.view_product.load_ms").build()
    }

    // MARK: - Product view

    /// 3-span tree + INFO log + histogram sample for a single product view.
    func emitProductView(product: Product, iteration: Int = 0) {
        let startNanos = DispatchTime.now().uptimeNanoseconds

        logger.logRecordBuilder()
            .setBody(.string("shop.view_product"))
            .setSeverity(.info)
            .setAttributes([
                "event.name": .string("shop.view_product"),
                "product.id": .string(product.id),
                "product.name": .string(product.name),
            ])
            .emit()

        let parent = tracer.spanBuilder(spanName: "shop.view_product")
            .setSpanKind(spanKind: .internal).startSpan()
        parent.setAttribute(key: "product.id", value: product.id)
        parent.setAttribute(key: "product.name", value: product.name)
        parent.setAttribute(key: "auto_demo.iteration", value: iteration)

        let reviews = tracer.spanBuilder(spanName: "shop.load_reviews")
            .setParent(parent).startSpan()
        reviews.setAttribute(key: "product.id", value: product.id)
        reviews.setAttribute(key: "reviews.count", value: Int.random(in: 3...42))
        reviews.status = .ok
        reviews.end()

        let recs = tracer.spanBuilder(spanName: "shop.load_recommendations")
            .setParent(parent).startSpan()
        recs.setAttribute(key: "recommendations.count", value: 4)
        recs.status = .ok
        recs.end()

        parent.status = .ok
        parent.end()

        let elapsedMs = Double(DispatchTime.now().uptimeNanoseconds - startNanos) / 1_000_000.0
        viewProductHistogram.record(
            value: elapsedMs,
            attributes: ["product.id": .string(product.id)]
        )
    }

    // MARK: - Cart

    /// Counter + INFO log (+ WARN log when `lineQuantity >= 5`) for a single
    /// cart-add action. Caller is responsible for updating cart state.
    ///
    /// - Parameters:
    ///   - quantityAdded: how many units this add operation added (counter delta)
    ///   - lineQuantity: cumulative units for the product in the cart after
    ///     the add — used to trigger the "large quantity" WARN log
    ///   - itemCount: total cart size after the add
    func emitCartAdd(
        product: Product,
        quantityAdded: Int,
        lineQuantity: Int,
        itemCount: Int
    ) {
        itemsAddedCounter.add(
            value: quantityAdded,
            attributes: [
                "product.id": .string(product.id),
                "product.category": .string(product.categories.first ?? "uncategorized"),
            ]
        )
        emitLog(event: "cart.add_item", attrs: [
            "product.id": .string(product.id),
            "product.name": .string(product.name),
            "cart.quantity": .int(quantityAdded),
            "cart.item_count": .int(itemCount),
        ])
        // Demo a WARN log path: large cumulative line quantity surfaces a
        // natural warn without needing a dedicated "press this to see a warn
        // log" button.
        if lineQuantity >= 5 {
            emitLog(event: "cart.large_quantity_warning", severity: .warn, attrs: [
                "product.id": .string(product.id),
                "cart.quantity": .int(lineQuantity),
            ])
        }
    }

    func emitCartRemove(product: Product, itemCount: Int) {
        emitLog(event: "cart.remove_item", attrs: [
            "product.id": .string(product.id),
            "cart.item_count": .int(itemCount),
        ])
    }

    func emitCartClear() {
        emitLog(event: "cart.cleared")
    }

    // MARK: - Checkout

    struct CheckoutLine {
        let productId: String
        let productName: String
        let quantity: Int
        let lineTotal: Double
    }

    /// 14-span deep nested checkout trace + histogram sample.
    ///
    /// Runs span emission on a detached Task so sleeps between spans don't
    /// block the caller's queue. The completion handler fires on the main
    /// queue so UI state updates are safe.
    func emitCheckout(
        lines: [CheckoutLine],
        totalUsd: Double,
        itemCount: Int,
        onComplete: @escaping @Sendable (Bool) -> Void
    ) {
        let startedMs = Date().timeIntervalSince1970 * 1000
        let parent = tracer.spanBuilder(spanName: "checkout")
            .setSpanKind(spanKind: .internal).startSpan()
        parent.setAttribute(key: "cart.item_count", value: itemCount)
        parent.setAttribute(key: "cart.line_count", value: lines.count)
        parent.setAttribute(key: "cart.total_usd", value: totalUsd)

        let tracer = self.tracer
        let histogram = self.checkoutDurationHistogram
        Task.detached {
            await Self.runCheckoutTrace(
                tracer: tracer, parent: parent, lines: lines, totalUsd: totalUsd
            )
            let durationMs = Date().timeIntervalSince1970 * 1000 - startedMs
            var histo = histogram
            histo.record(
                value: durationMs,
                attributes: [
                    "checkout.item_count": .int(itemCount),
                    "checkout.line_count": .int(lines.count),
                ]
            )
            DispatchQueue.main.async { onComplete(true) }
        }
    }

    // MARK: - Internal helpers

    private func emitLog(
        event: String,
        severity: Severity = .info,
        attrs: [String: AttributeValue] = [:]
    ) {
        var a = attrs
        a["event.name"] = .string(event)
        logger.logRecordBuilder()
            .setBody(.string(event))
            .setSeverity(severity)
            .setAttributes(a)
            .emit()
    }

    /// Shared nested-span emission. Static + nonisolated so sleeps between
    /// spans reflect real work, not main-thread scheduling.
    private static func runCheckoutTrace(
        tracer: Tracer,
        parent: Span,
        lines: [CheckoutLine],
        totalUsd: Double
    ) async {
        // 1. Validate cart
        let validate = tracer.spanBuilder(spanName: "checkout.validate_cart")
            .setParent(parent).startSpan()
        validate.setAttribute(key: "cart.line_count", value: lines.count)
        try? await Task.sleep(nanoseconds: 40_000_000)
        validate.status = .ok; validate.end()

        // 2. Inventory check: parent + per-line child spans (cap at 4 for trace size)
        let inventory = tracer.spanBuilder(spanName: "checkout.inventory_check")
            .setParent(parent).startSpan()
        inventory.setAttribute(key: "inventory.line_count", value: lines.count)
        for line in lines.prefix(4) {
            let item = tracer.spanBuilder(spanName: "inventory.check_item")
                .setParent(inventory).startSpan()
            item.setAttribute(key: "product.id", value: line.productId)
            item.setAttribute(key: "cart.quantity", value: line.quantity)
            try? await Task.sleep(nanoseconds: 12_000_000)
            item.setAttribute(key: "inventory.in_stock", value: true)
            item.status = .ok; item.end()
        }
        inventory.status = .ok; inventory.end()

        // 3. Calculate totals: parent + 3 children (subtotal, tax, shipping)
        let totals = tracer.spanBuilder(spanName: "checkout.calculate_totals")
            .setParent(parent).startSpan()
        let subtotal = tracer.spanBuilder(spanName: "totals.subtotal")
            .setParent(totals).startSpan()
        subtotal.setAttribute(key: "totals.subtotal_usd", value: totalUsd)
        try? await Task.sleep(nanoseconds: 10_000_000)
        subtotal.status = .ok; subtotal.end()

        let tax = tracer.spanBuilder(spanName: "totals.tax")
            .setParent(totals).startSpan()
        let taxUsd = totalUsd * 0.0825
        tax.setAttribute(key: "totals.tax_usd", value: taxUsd)
        try? await Task.sleep(nanoseconds: 10_000_000)
        tax.status = .ok; tax.end()

        let shipping = tracer.spanBuilder(spanName: "totals.shipping")
            .setParent(totals).startSpan()
        let shippingUsd: Double = totalUsd > 50 ? 0 : 5.99
        shipping.setAttribute(key: "totals.shipping_usd", value: shippingUsd)
        shipping.setAttribute(key: "totals.free_shipping", value: totalUsd > 50)
        try? await Task.sleep(nanoseconds: 10_000_000)
        shipping.status = .ok; shipping.end()

        totals.setAttribute(key: "totals.grand_total_usd", value: totalUsd + taxUsd + shippingUsd)
        totals.status = .ok; totals.end()

        // 4. Charge: parent + validate-card + authorize
        let charge = tracer.spanBuilder(spanName: "checkout.charge")
            .setParent(parent).startSpan()
        charge.setAttribute(key: "payment.method", value: "card")
        let validateCard = tracer.spanBuilder(spanName: "payment.validate_card")
            .setParent(charge).startSpan()
        try? await Task.sleep(nanoseconds: 40_000_000)
        validateCard.status = .ok; validateCard.end()
        let authorize = tracer.spanBuilder(spanName: "payment.authorize")
            .setParent(charge).startSpan()
        authorize.setAttribute(key: "payment.amount_usd", value: totalUsd + taxUsd + shippingUsd)
        try? await Task.sleep(nanoseconds: 120_000_000)
        authorize.status = .ok; authorize.end()
        charge.status = .ok; charge.end()

        // 5. Send confirmation (email render + send)
        let confirm = tracer.spanBuilder(spanName: "checkout.send_confirmation")
            .setParent(parent).startSpan()
        let render = tracer.spanBuilder(spanName: "email.render")
            .setParent(confirm).startSpan()
        try? await Task.sleep(nanoseconds: 20_000_000)
        render.status = .ok; render.end()
        let send = tracer.spanBuilder(spanName: "email.send")
            .setParent(confirm).startSpan()
        send.setAttribute(key: "email.provider", value: "demo-mailer")
        try? await Task.sleep(nanoseconds: 30_000_000)
        send.status = .ok; send.end()
        confirm.status = .ok; confirm.end()

        // 6. Analytics report
        let analytics = tracer.spanBuilder(spanName: "checkout.analytics.report")
            .setParent(parent).startSpan()
        analytics.setAttribute(key: "analytics.provider", value: "dash0")
        try? await Task.sleep(nanoseconds: 15_000_000)
        analytics.status = .ok; analytics.end()

        parent.status = .ok; parent.end()
    }
}

File: examples/upstream-demo-app/src/main/java/io/opentelemetry/android/demo/shop/ShopTelemetry.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.demo.shop

import io.opentelemetry.android.demo.OtelDemoApplication
import io.opentelemetry.android.demo.shop.model.Product
import io.opentelemetry.api.common.AttributeKey.doubleKey
import io.opentelemetry.api.common.AttributeKey.longKey
import io.opentelemetry.api.common.AttributeKey.stringKey
import io.opentelemetry.api.logs.Severity
import io.opentelemetry.api.metrics.DoubleHistogram
import io.opentelemetry.api.metrics.LongCounter
import io.opentelemetry.api.trace.Span
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.context.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Single source of truth for every shop log, span, and metric this app emits.
 * Mirrors [examples/upstream-demo-app-ios/AstronomyShop/Shop/ShopTelemetry.swift]
 * on iOS — the contract is documented in docs/design/shop-telemetry-contract.md.
 *
 * Any drift between this file and its iOS twin is a cross-platform parity
 * bug and must be fixed in the same commit that introduces it.
 *
 * The class is a plain Kotlin object so any call site (UI composables,
 * view models, the auto-demo driver) can share the same instrument
 * handles. The OTel tracer / logger / meter are fetched lazily from
 * `OtelDemoApplication.openTelemetry` — callsites that fire before SDK
 * init get a silent no-op, same as the iOS side when `OTelMobile.start`
 * hasn't completed yet.
 */
object ShopTelemetry {
    private const val INSTRUMENTATION_SCOPE = "io.dash0.demo.shop"

    private val tracer: Tracer?
        get() = OtelDemoApplication.openTelemetry?.getTracer(INSTRUMENTATION_SCOPE)

    private val logger by lazy {
        OtelDemoApplication.openTelemetry?.logsBridge?.get(INSTRUMENTATION_SCOPE)
    }

    private val meter by lazy {
        OtelDemoApplication.openTelemetry?.getMeter(INSTRUMENTATION_SCOPE)
    }

    private val itemsAddedCounter: LongCounter? by lazy {
        meter?.counterBuilder("shop.cart.items_added")?.build()
    }

    private val checkoutDurationHistogram: DoubleHistogram? by lazy {
        meter?.histogramBuilder("shop.checkout.duration_ms")?.build()
    }

    private val viewProductHistogram: DoubleHistogram? by lazy {
        meter?.histogramBuilder("shop.view_product.load_ms")?.build()
    }

    // ------------------------------------------------------------------
    // Logs
    // ------------------------------------------------------------------

    fun emitAppHomeAppeared() {
        emitLog("app.home_appeared", Severity.INFO)
    }

    fun emitProductView(product: Product, iteration: Int = 0) {
        val startNanos = System.nanoTime()

        emitLog(
            "shop.view_product", Severity.INFO,
            attrs = mapOf(
                "product.id" to product.id,
                "product.name" to product.name
            )
        )

        val tr = tracer
        if (tr != null) {
            val parent = tr.spanBuilder("shop.view_product")
                .setSpanKind(SpanKind.INTERNAL)
                .setAttribute("product.id", product.id)
                .setAttribute("product.name", product.name)
                .setAttribute("auto_demo.iteration", iteration.toLong())
                .startSpan()
            val parentCtx = Context.current().with(parent)

            tr.spanBuilder("shop.load_reviews")
                .setParent(parentCtx)
                .setAttribute("product.id", product.id)
                .setAttribute("reviews.count", (3L + (iteration * 7 % 40)))
                .startSpan()
                .apply {
                    setStatus(StatusCode.OK)
                    end()
                }

            tr.spanBuilder("shop.load_recommendations")
                .setParent(parentCtx)
                .setAttribute("recommendations.count", 4L)
                .startSpan()
                .apply {
                    setStatus(StatusCode.OK)
                    end()
                }

            parent.setStatus(StatusCode.OK)
            parent.end()
        }

        val elapsedMs = (System.nanoTime() - startNanos) / 1_000_000.0
        viewProductHistogram?.record(
            elapsedMs,
            io.opentelemetry.api.common.Attributes.of(stringKey("product.id"), product.id)
        )
    }

    fun emitCartAdd(product: Product, quantityAdded: Int, lineQuantity: Int, itemCount: Int) {
        itemsAddedCounter?.add(
            quantityAdded.toLong(),
            io.opentelemetry.api.common.Attributes.of(
                stringKey("product.id"), product.id,
                stringKey("product.category"), product.categories.firstOrNull() ?: "uncategorized"
            )
        )
        emitLog(
            "cart.add_item", Severity.INFO,
            attrs = mapOf(
                "product.id" to product.id,
                "product.name" to product.name,
                "cart.quantity" to quantityAdded.toLong(),
                "cart.item_count" to itemCount.toLong()
            )
        )
        if (lineQuantity >= 5) {
            emitLog(
                "cart.large_quantity_warning", Severity.WARN,
                attrs = mapOf(
                    "product.id" to product.id,
                    "cart.quantity" to lineQuantity.toLong()
                )
            )
        }
    }

    fun emitCartRemove(product: Product, itemCount: Int) {
        emitLog(
            "cart.remove_item", Severity.INFO,
            attrs = mapOf(
                "product.id" to product.id,
                "cart.item_count" to itemCount.toLong()
            )
        )
    }

    fun emitCartClear() {
        emitLog("cart.cleared", Severity.INFO)
    }

    // ------------------------------------------------------------------
    // Checkout — 14-span trace + histogram sample
    // ------------------------------------------------------------------

    data class CheckoutLine(
        val productId: String,
        val productName: String,
        val quantity: Int,
        val lineTotal: Double
    )

    fun emitCheckout(
        lines: List<CheckoutLine>,
        totalUsd: Double,
        itemCount: Int,
        scope: CoroutineScope,
        onComplete: (Boolean) -> Unit = {}
    ): Job? {
        val tr = tracer ?: run { onComplete(false); return null }
        val startedMs = System.currentTimeMillis()

        val parent = tr.spanBuilder("checkout")
            .setSpanKind(SpanKind.INTERNAL)
            .setAttribute("cart.item_count", itemCount.toLong())
            .setAttribute("cart.line_count", lines.size.toLong())
            .setAttribute("cart.total_usd", totalUsd)
            .startSpan()

        return scope.launch(Dispatchers.Default) {
            runCheckoutTrace(tr, parent, lines, totalUsd)
            val durationMs = (System.currentTimeMillis() - startedMs).toDouble()
            checkoutDurationHistogram?.record(
                durationMs,
                io.opentelemetry.api.common.Attributes.of(
                    longKey("checkout.item_count"), itemCount.toLong(),
                    longKey("checkout.line_count"), lines.size.toLong()
                )
            )
            onComplete(true)
        }
    }

    private suspend fun runCheckoutTrace(
        tracer: Tracer,
        parent: Span,
        lines: List<CheckoutLine>,
        totalUsd: Double
    ) {
        val parentCtx = Context.current().with(parent)

        // 1. Validate cart
        childSpan(tracer, parentCtx, "checkout.validate_cart") { s ->
            s.setAttribute("cart.line_count", lines.size.toLong())
            delay(40)
        }

        // 2. Inventory check — parent + per-line children (cap 4)
        val inventory = tracer.spanBuilder("checkout.inventory_check")
            .setParent(parentCtx).startSpan()
        inventory.setAttribute("inventory.line_count", lines.size.toLong())
        val inventoryCtx = Context.current().with(inventory)
        for (line in lines.take(4)) {
            childSpan(tracer, inventoryCtx, "inventory.check_item") { s ->
                s.setAttribute("product.id", line.productId)
                s.setAttribute("cart.quantity", line.quantity.toLong())
                s.setAttribute("inventory.in_stock", true)
                delay(12)
            }
        }
        inventory.setStatus(StatusCode.OK); inventory.end()

        // 3. Calculate totals — parent + 3 children
        val totals = tracer.spanBuilder("checkout.calculate_totals")
            .setParent(parentCtx).startSpan()
        val totalsCtx = Context.current().with(totals)
        val taxUsd = totalUsd * 0.0825
        val shippingUsd = if (totalUsd > 50) 0.0 else 5.99

        childSpan(tracer, totalsCtx, "totals.subtotal") { s ->
            s.setAttribute("totals.subtotal_usd", totalUsd); delay(10)
        }
        childSpan(tracer, totalsCtx, "totals.tax") { s ->
            s.setAttribute("totals.tax_usd", taxUsd); delay(10)
        }
        childSpan(tracer, totalsCtx, "totals.shipping") { s ->
            s.setAttribute("totals.shipping_usd", shippingUsd)
            s.setAttribute("totals.free_shipping", totalUsd > 50)
            delay(10)
        }

        totals.setAttribute("totals.grand_total_usd", totalUsd + taxUsd + shippingUsd)
        totals.setStatus(StatusCode.OK); totals.end()

        // 4. Charge — parent + validate-card + authorize
        val charge = tracer.spanBuilder("checkout.charge")
            .setParent(parentCtx).startSpan()
        charge.setAttribute("payment.method", "card")
        val chargeCtx = Context.current().with(charge)

        childSpan(tracer, chargeCtx, "payment.validate_card") { _ -> delay(40) }
        childSpan(tracer, chargeCtx, "payment.authorize") { s ->
            s.setAttribute("payment.amount_usd", totalUsd + taxUsd + shippingUsd)
            delay(120)
        }
        charge.setStatus(StatusCode.OK); charge.end()

        // 5. Send confirmation — email render + send
        val confirm = tracer.spanBuilder("checkout.send_confirmation")
            .setParent(parentCtx).startSpan()
        val confirmCtx = Context.current().with(confirm)

        childSpan(tracer, confirmCtx, "email.render") { _ -> delay(20) }
        childSpan(tracer, confirmCtx, "email.send") { s ->
            s.setAttribute("email.provider", "demo-mailer"); delay(30)
        }
        confirm.setStatus(StatusCode.OK); confirm.end()

        // 6. Analytics report
        childSpan(tracer, parentCtx, "checkout.analytics.report") { s ->
            s.setAttribute("analytics.provider", "dash0"); delay(15)
        }

        parent.setStatus(StatusCode.OK); parent.end()
    }

    private suspend fun childSpan(
        tracer: Tracer,
        parentCtx: Context,
        name: String,
        block: suspend (Span) -> Unit
    ) {
        val span = tracer.spanBuilder(name).setParent(parentCtx).startSpan()
        try {
            block(span)
            span.setStatus(StatusCode.OK)
        } finally {
            span.end()
        }
    }

    // ------------------------------------------------------------------
    // Internal log helper — puts `event.name` both on the body and on an
    // attribute so consumers that filter either way work.
    // ------------------------------------------------------------------

    private fun emitLog(
        eventName: String,
        severity: Severity,
        attrs: Map<String, Any> = emptyMap()
    ) {
        val l = logger ?: return
        val builder = l.logRecordBuilder()
            .setBody(eventName)
            .setSeverity(severity)
            .setAttribute(stringKey("event.name"), eventName)
        for ((key, value) in attrs) {
            when (value) {
                is String -> builder.setAttribute(stringKey(key), value)
                is Long -> builder.setAttribute(longKey(key), value)
                is Int -> builder.setAttribute(longKey(key), value.toLong())
                is Double -> builder.setAttribute(doubleKey(key), value)
                is Float -> builder.setAttribute(doubleKey(key), value.toDouble())
                is Boolean -> builder.setAttribute(
                    io.opentelemetry.api.common.AttributeKey.booleanKey(key),
                    value
                )
                else -> builder.setAttribute(stringKey(key), value.toString())
            }
        }
        builder.emit()
    }
}
